drivers/char/tpm/tpm_ftpm_optee.c

```diff
@@ -0,0 +1,304 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ *
+ * Author: Jiri Appl <jiria@microsoft.com>
+ *
+ * OpTEE fTPM TA TPM driver
+ * 
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, version 2 of the
+ * License.
+ *
+ */
+
+#include <linux/tee_drv.h>
+
+#include "tpm.h"
+
+#define TEEC_CONFIG_PAYLOAD_REF_COUNT 4
+#define TEEC_LOGIN_PUBLIC             0x00000000
+#define TEEC_SUCCESS                  0x00000000
+
+#define FTPM_OPTEE_TA_SUBMIT_COMMAND  0
+
+struct tpm_tee_context 
+{
+	struct tpm_chip *chip;
+
+	size_t response_length;
+	struct mutex response_buffer_lock;
+	u8 response_buffer[TPM_BUFSIZE];
+
+	struct tee_context *ftpm_optee_context;
+
+	bool optee_session_opened;
+	u32 optee_session_id;
+
+	bool chip_registered;
+};
+
+static struct tpm_tee_context tpm_tee_context = { 0 };
+
+static int ftpm_optee_tpm_op_recv(struct tpm_chip *chip, u8 *buf, size_t len)
+{
+	size_t response_length;
+	int rc;
+
+	mutex_lock(&tpm_tee_context.response_buffer_lock);
+
+	response_length = tpm_tee_context.response_length;
+	if (len < response_length) 
+	{
+		rc = -EIO;
+		goto cleanup;
+	}
+
+	memcpy(buf, tpm_tee_context.response_buffer, response_length);
+	tpm_tee_context.response_length = 0;
+
+	rc = response_length;
+
+cleanup:
+	mutex_unlock(&tpm_tee_context.response_buffer_lock);
+	return rc;
+}
+
+static int ftpm_optee_tpm_op_send(struct tpm_chip *chip, u8 *buf, size_t len)
+{
+	int rc = 0;
+	size_t response_length;
+	u8* response_buffer;
+	struct tpm_output_header* response_header;
+	struct tee_param submit_command_params[TEEC_CONFIG_PAYLOAD_REF_COUNT] = {
+		{ .attr = TEE_IOCTL_PARAM_ATTR_TYPE_MEMREF_INPUT },
+		{ .attr = TEE_IOCTL_PARAM_ATTR_TYPE_MEMREF_INOUT },
+		{ .attr = TEE_IOCTL_PARAM_ATTR_TYPE_NONE },
+		{ .attr = TEE_IOCTL_PARAM_ATTR_TYPE_NONE }
+	};
+	struct tee_ioctl_invoke_arg submit_command_args;
+	struct tee_shm *shm = NULL;
+
+	mutex_lock(&tpm_tee_context.response_buffer_lock);
+
+	tpm_tee_context.response_length = 0;
+
+	submit_command_args.func = FTPM_OPTEE_TA_SUBMIT_COMMAND;
+	submit_command_args.session = tpm_tee_context.optee_session_id;
+	submit_command_args.num_params = TEEC_CONFIG_PAYLOAD_REF_COUNT;
+
+	shm = tee_shm_alloc(
+		tpm_tee_context.ftpm_optee_context, 
+		len + sizeof(tpm_tee_context.response_buffer), 
+		TEE_SHM_MAPPED | TEE_SHM_DMA_BUF);
```

Comment:
My understanding is that DMA buffers are used to allocate shared memory mapped in user-space. Is `TEE_SHM_DMA_BUF` required here?

Reply:
The alloc would fail without the TEE_SHM_DMA_BUF flag. I did not dig further. If you feel like we should investigate more, I will put it on my TODO list.

Reply:
Thinking more about this, this might have something to do with the TEEC patches I picked up. They had bunch of dependent patches missing in the NXP Linux fork and so it is possible that I missed another or merged something incorrectly. I will keep this on my radar and will come back to this after 15th.

Reply:
Thanks for looking!